// @ricky0123/vad-react/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useMicVAD = exports.getDefaultReactRealTimeVADOptions = exports.utils = void 0;
const vad_web_1 = require("@ricky0123/vad-web");
const react_1 = require("react");
var vad_web_2 = require("@ricky0123/vad-web");
Object.defineProperty(exports, "utils", { enumerable: true, get: function () { return vad_web_2.utils; } });
const defaultReactOptions = {
    userSpeakingThreshold: 0.6,
};
const getDefaultReactRealTimeVADOptions = (model) => {
    return {
        ...(0, vad_web_1.getDefaultRealTimeVADOptions)(model),
        ...defaultReactOptions,
    };
};
exports.getDefaultReactRealTimeVADOptions = getDefaultReactRealTimeVADOptions;
function useOptions(options) {
    const model = options.model ?? vad_web_1.DEFAULT_MODEL;
    const fullOptions = {
        ...(0, exports.getDefaultReactRealTimeVADOptions)(model),
        ...options,
    };
    const reactOptions = {
        userSpeakingThreshold: fullOptions.userSpeakingThreshold,
    };
    const vadOptions = {
        positiveSpeechThreshold: fullOptions.positiveSpeechThreshold,
        negativeSpeechThreshold: fullOptions.negativeSpeechThreshold,
        redemptionMs: fullOptions.redemptionMs,
        preSpeechPadMs: fullOptions.preSpeechPadMs,
        minSpeechMs: fullOptions.minSpeechMs,
        submitUserSpeechOnPause: fullOptions.submitUserSpeechOnPause,
        onFrameProcessed: fullOptions.onFrameProcessed,
        onVADMisfire: fullOptions.onVADMisfire,
        onSpeechStart: fullOptions.onSpeechStart,
        onSpeechEnd: fullOptions.onSpeechEnd,
        onSpeechRealStart: fullOptions.onSpeechRealStart,
        baseAssetPath: fullOptions.baseAssetPath,
        onnxWASMBasePath: fullOptions.onnxWASMBasePath,
        model: fullOptions.model,
        workletOptions: fullOptions.workletOptions,
        getStream: fullOptions.getStream,
        pauseStream: fullOptions.pauseStream,
        resumeStream: fullOptions.resumeStream,
        startOnLoad: fullOptions.startOnLoad,
        processorType: fullOptions.processorType,
    };
    if (fullOptions.ortConfig) {
        vadOptions.ortConfig = fullOptions.ortConfig;
    }
    if (fullOptions.audioContext) {
        vadOptions.audioContext = fullOptions.audioContext;
    }
    return [reactOptions, vadOptions];
}
function useMicVAD(options) {
    const [reactOptions, vadOptions] = useOptions(options);
    const model = options["model"] ?? vad_web_1.DEFAULT_MODEL;
    const [userSpeaking, updateUserSpeaking] = (0, react_1.useState)(false);
    const [loading, setLoading] = (0, react_1.useState)(true);
    const [errored, setErrored] = (0, react_1.useState)(false);
    const [listening, setListening] = (0, react_1.useState)(false);
    const [vad, setVAD] = (0, react_1.useState)(null);
    // Use refs to store the latest callbacks so they can be called without recreating the VAD
    const onFrameProcessedRef = (0, react_1.useRef)(vadOptions.onFrameProcessed);
    const onSpeechEndRef = (0, react_1.useRef)(vadOptions.onSpeechEnd);
    const onSpeechStartRef = (0, react_1.useRef)(vadOptions.onSpeechStart);
    const onSpeechRealStartRef = (0, react_1.useRef)(vadOptions.onSpeechRealStart);
    const onVADMisfireRef = (0, react_1.useRef)(vadOptions.onVADMisfire);
    const getStreamRef = (0, react_1.useRef)(vadOptions.getStream);
    // Update refs when callbacks change
    (0, react_1.useEffect)(() => {
        onFrameProcessedRef.current = vadOptions.onFrameProcessed;
        onSpeechEndRef.current = vadOptions.onSpeechEnd;
        onSpeechStartRef.current = vadOptions.onSpeechStart;
        onSpeechRealStartRef.current = vadOptions.onSpeechRealStart;
        onVADMisfireRef.current = vadOptions.onVADMisfire;
    }, [
        vadOptions.onFrameProcessed,
        vadOptions.onSpeechEnd,
        vadOptions.onSpeechStart,
        vadOptions.onSpeechRealStart,
        vadOptions.onVADMisfire,
    ]);
    // Update getStream ref - this is the key fix!
    (0, react_1.useEffect)(() => {
        getStreamRef.current = vadOptions.getStream;
    }, [vadOptions.getStream]);
    // Serialize getStream function to detect changes
    // We use a simple approach: convert function to string
    const getStreamKey = vadOptions.getStream.toString();
    (0, react_1.useEffect)(() => {
        let myvad = null;
        let canceled = false;
        const setup = async () => {
            try {
                setLoading(true);
                setErrored(false);
                // Create VAD options with stable callback wrappers
                const finalVadOptions = {
                    ...vadOptions,
                    onFrameProcessed: (probs, frame) => {
                        const isSpeaking = probs.isSpeech > reactOptions.userSpeakingThreshold;
                        updateUserSpeaking(isSpeaking);
                        void onFrameProcessedRef.current(probs, frame);
                    },
                    onSpeechEnd: (audio) => {
                        void onSpeechEndRef.current(audio);
                    },
                    onSpeechStart: () => {
                        void onSpeechStartRef.current();
                    },
                    onSpeechRealStart: () => {
                        void onSpeechRealStartRef.current();
                    },
                    onVADMisfire: () => {
                        void onVADMisfireRef.current();
                    },
                    getStream: () => {
                        return getStreamRef.current();
                    },
                };
                myvad = await vad_web_1.MicVAD.new(finalVadOptions);
                if (canceled) {
                    await myvad.destroy();
                    return;
                }
                setVAD(myvad);
                if (vadOptions.startOnLoad) {
                    await myvad.start();
                    setListening(true);
                }
                setLoading(false);
            }
            catch (e) {
                setLoading(false);
                if (e instanceof Error) {
                    setErrored(e.message);
                }
                else {
                    setErrored(String(e));
                }
            }
        };
        setup().catch(() => {
            // Error already handled in setup function
        });
        return function cleanUp() {
            canceled = true;
            if (myvad) {
                void myvad.destroy();
            }
            if (!loading && !errored) {
                setListening(false);
            }
        };
    }, [getStreamKey, model]); // Recreate when getStream changes or model changes
    const pause = (0, react_1.useCallback)(async () => {
        if (!loading && !errored) {
            await vad?.pause();
            setListening(false);
        }
    }, [loading, errored, vad]);
    const start = (0, react_1.useCallback)(async () => {
        if (!loading && !errored) {
            await vad?.start();
            setListening(true);
        }
    }, [loading, errored, vad]);
    const toggle = (0, react_1.useCallback)(async () => {
        if (listening) {
            await pause();
        }
        else {
            await start();
        }
    }, [listening, pause, start]);
    return {
        listening,
        errored,
        loading,
        userSpeaking,
        pause,
        start,
        toggle,
    };
}
exports.useMicVAD = useMicVAD;
